homeassistant/components/calendar/__init__.py

```diff
@@ -250,6 +257,20 @@ def _validate_rrule(value: Any) -> str:
     extra=vol.ALLOW_EXTRA,
 )
 
+SERVICE_LIST_EVENTS: Final = "list_events"
+SERVICE_LIST_EVENTS_SCHEMA: Final = vol.All(
+    cv.has_at_most_one_key(EVENT_END_DATETIME, EVENT_DURATION),
+    cv.make_entity_service_schema(
+        {
+            vol.Optional(EVENT_START_DATETIME): datetime.datetime,
+            vol.Optional(EVENT_END_DATETIME): datetime.datetime,
+            vol.Optional(EVENT_DURATION): vol.All(
```

Comment:
Use vol.Exclusive here and group duration and eg start time.

Reply:
Probably use two schemas and combine them with `vol.All` to gather all the requirements.

Reply:
Can you elaborate?  My understanding of this is:
- start is optional
- either duration or end are required, but not both (has at most one key, at least one key)

I think understand the /what/, but not the /why/ and the higher level goal of the suggestion. Are we missing some test cases for failures? or is this simpler code better than the lower level constructs? Is this suggesting a replacement or complementary to what is already here? Generally, while voluptuous has a lot of examples in documentation, I think its missing a user guide especially for more advanced things.

Reply:
These are the rules I suggest:
If duration is set, start and end should not be set. If start and end is set, duration should not be set. If any of start or end is set, both need to be set. Either duration, or start and end must be set.

To accomplish this we can use `cv.has_at_least_one_key` for duration and one of start or end, `vol.Exclusive` for duration and one of start or end, and `vol.Inclusive` for start and end.

Reply:
Let's drop the exclusive rule. It's enough to just make sure not both duration and end is set.